### PR TITLE
[prometheus-push-gateway] Add lifecycle hooks

### LINKED ISSUE #5102: [prometheus-pushgateway] Support lifecycle hooks
### Is your feature request related to a problem ?

I'm not able to add a pre-stop hook to the push-gateway deployment.

### Describe the solution you'd like.

I would like to be able to add a sleep to the pods before being terminated.

### Describe alternatives you've considered.

I can manually add this to the deployment, but I would like to keep this in source control.

### Additional context.

_No response_

### PR DESCRIPTION
**Pull request by duncaan**
Fixes #5102

Add support for lifecycle hooks in the push-gateway deployment.

* Add lifecycle hooks configuration to `charts/prometheus-pushgateway/templates/deployment.yaml`
* Add lifecycle hooks configuration to `charts/prometheus-pushgateway/templates/statefulset.yaml`
* Add lifecycle hooks configuration options in `charts/prometheus-pushgateway/values.yaml`

---

For more details, open the [Copilot Workspace session](https://copilot-workspace.githubnext.com/prometheus-community/helm-charts/pull/5103?shareId=0d96fb60-61f4-458c-ad32-e5e0e0c1a16e).

### CODE REVIEW

File: charts/prometheus-pushgateway/templates/deployment.yaml [hunk 1]
```diff
@@ -25,4 +25,6 @@ spec:
         {{- include "prometheus-pushgateway.defaultLabels" . | nindent 8 }}
     spec:
       {{- include "prometheus-pushgateway.podSpec" . | nindent 6 }}
+      lifecycle:
+        {{- toYaml .Values.lifecycle | nindent 8 }}
```
Comment (zeritti) [inline charts/prometheus-pushgateway/templates/deployment.yaml:29]: I'd suggest not to set an empty field:
```suggestion
      {{- with .Values.lifecycle }}
      lifecycle:
        {{- toYaml . | nindent 8 }}
      {{- end }}
```

File: charts/prometheus-pushgateway/templates/statefulset.yaml [hunk 2]
```diff
@@ -22,6 +22,8 @@ spec:
         {{- include "prometheus-pushgateway.defaultLabels" . | nindent 8 }}
     spec:
       {{- include "prometheus-pushgateway.podSpec" . | nindent 6 }}
+      lifecycle:
+        {{- toYaml .Values.lifecycle | nindent 8 }}
```
Comment (zeritti) [inline charts/prometheus-pushgateway/templates/statefulset.yaml:26]: As above.

**Review by zeritti — CHANGES_REQUESTED** [review]
Thank you, @duncaan, for your PR. Please, see my comments below.

File: charts/prometheus-pushgateway/templates/deployment.yaml (hunk 1 above)
Comment (duncaan) [reply charts/prometheus-pushgateway/templates/deployment.yaml]: Thank you for the feedback.

**Review by zeritti — CHANGES_REQUESTED** [review]
Thanks for your changes. I did not immediately realise then that podSpec was defined in the helper template `prometheus-pushgateway.podSpec` in `_helpers.tpl`. The lifecycle field has to go in there.

**Review by zeritti — APPROVED** [review]
Placement of the lifecycle field has been corrected, a test case is included.

**Pull request merged**